futures-util/src/try_stream/try_filter.rs

```diff
@@ -0,0 +1,118 @@
+use core::pin::Pin;
+use futures_core::future::Future;
+use futures_core::stream::{Stream, TryStream};
+use futures_core::task::{Context, Poll};
+use futures_sink::Sink;
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+
+/// Stream for the [`try_filter`](super::TryStreamExt::try_filter)
+/// method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct TryFilter<St, Fut, F>
+    where St: TryStream
+{
+    stream: St,
+    f: F,
+    pending_fut: Option<Fut>,
+    pending_item: Option<St::Ok>,
+}
+
+impl<St, Fut, F> Unpin for TryFilter<St, Fut, F>
+    where St: TryStream + Unpin, Fut: Unpin,
+{}
+
+impl<St, Fut, F> TryFilter<St, Fut, F>
+    where St: TryStream
+{
+    unsafe_pinned!(stream: St);
+    unsafe_unpinned!(f: F);
+    unsafe_pinned!(pending_fut: Option<Fut>);
+    unsafe_unpinned!(pending_item: Option<St::Ok>);
+
+    pub(super) fn new(stream: St, f: F) -> Self {
+        TryFilter {
+            stream,
+            f,
+            pending_fut: None,
+            pending_item: None,
+        }
+    }
+
+    /// Acquires a reference to the underlying stream that this combinator is
+    /// pulling from.
+    pub fn get_ref(&self) -> &St {
+        &self.stream
+    }
+
+    /// Acquires a mutable reference to the underlying stream that this
+    /// combinator is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the
+    /// stream which may otherwise confuse this combinator.
+    pub fn get_mut(&mut self) -> &mut St {
+        &mut self.stream
+    }
+
+    /// Acquires a pinned mutable reference to the underlying stream that this
+    /// combinator is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the
+    /// stream which may otherwise confuse this combinator.
+    pub fn get_pin_mut<'a>(self: Pin<&'a mut Self>) -> Pin<&'a mut St> {
+        unsafe { Pin::map_unchecked_mut(self, Self::get_mut) }
```

Comment:
This should just call `self.stream()` to avoid the `unsafe` block.

Reply:
This one was copied from the [TryFilterMap::get_pin_mut](https://github.com/rust-lang-nursery/futures-rs/blob/5a6b1b32329b3f4a0ad1df3df1576fb36fdc5361/futures-util/src/try_stream/try_filter_map.rs#L46-L53) as is.

Maybe @taiki-e can help us here?

Reply:
It is preferable to replace the unsafe code with the safe code.
```diff
- unsafe { Pin::map_unchecked_mut(self, Self::get_mut) }
+ self.stream()
```
(The other combinators are processed by https://github.com/rust-lang-nursery/futures-rs/pull/1559.)